src/app/Http/Controllers/CrudController.php

```diff
@@ -38,24 +37,53 @@ public function __construct()
         // It's done inside a middleware closure in order to have
         // the complete request inside the CrudPanel object.
         $this->middleware(function ($request, $next) {
-            $this->crud = app('crud');
+            if (! Backpack::hasCrudController(get_class($this))) {
+                $this->initializeCrud($request);
 
-            $this->crud->setRequest($request);
+                return $next($request);
+            }
 
-            LifecycleHook::trigger('crud:before_setup_defaults', [$this]);
-            $this->setupDefaults();
-            LifecycleHook::trigger('crud:after_setup_defaults', [$this]);
+            $this->triggerControllerHooks();
 
-            LifecycleHook::trigger('crud:before_setup', [$this]);
-            $this->setup();
-            LifecycleHook::trigger('crud:after_setup', [$this]);
+            $this->initialized = true;
 
-            $this->setupConfigurationForCurrentOperation();
+            Backpack::crud($this)->setRequest($request);
 
             return $next($request);
         });
     }
 
+    public function initializeCrud($request, $crudPanel = null, $operation = null): void
+    {
+        $crudPanel ??= Backpack::crud($this);
+
+        if ($crudPanel->isInitialized()) {
+            $crudPanel->setRequest($request);
+
+            Backpack::setControllerCrud(get_class($this), $crudPanel);
+        }
+
+        $crudPanel->initialized = true;
+        $crudPanel->setRequest($request);
+
+        $this->triggerControllerHooks();
+
+        Backpack::setControllerCrud(get_class($this), $crudPanel);
+    }
+
+    private function triggerControllerHooks()
```

Comment:
I think this is a bad name for this function. Sure, it triggers the controller hooks... but that's just the SMALL part of what it does. What it does MORE IMPORTANTLY is to... setup the crud controller? O what is it? Cause I see it's calling the general-purpose setup methods like `setupDefaults()` and `setup()` but it's also calling `setupConfigurationForCurrentOperation()` which is an operation-thing.

Perhaps we call this method:
```suggestion
    private function setupCrudController()
```

If you don't have a better name for it.